sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/TDigestQuantiles.java

```diff
@@ -0,0 +1,321 @@
+   * at a given quantile or the approximate quantile position of a given element.
+   */
+  public static GlobalDigest globally() {
+    return GlobalDigest.builder().build();
+  }
+
+  /**
+   * Like {@link #globally()}, but builds a digest for each key in the stream.
+   *
+   * @param <K>             the type of the keys
+   */
+  public static <K> PerKeyDigest<K> perKey() {
+    return PerKeyDigest.<K>builder().build();
+  }
+
+  /** Implementation of {@link #globally()}. */
+  @AutoValue
+  public abstract static class GlobalDigest
+          extends PTransform<PCollection<Double>, PCollection<MergingDigest>> {
+
+    abstract double compression();
+
+    abstract Builder toBuilder();
+
+    static Builder builder() {
+      return new AutoValue_TDigestQuantiles_GlobalDigest.Builder()
+              .setCompression(100);
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setCompression(double cf);
+
+      abstract GlobalDigest build();
+    }
+
+    public GlobalDigest withCompression(double cf) {
+      return toBuilder().setCompression(cf).build();
+    }
+
+    @Override
+    public PCollection<MergingDigest> expand(PCollection<Double> input) {
+      return input.apply(
+                      "Compute HyperLogLog Structure",
+                      Combine.globally(TDigestQuantilesFn.create(this.compression())));
+    }
+  }
+
+  /** Implementation of {@link #perKey()}. */
+  @AutoValue
+  public abstract static class PerKeyDigest<K>
+          extends PTransform<PCollection<KV<K, Double>>, PCollection<KV<K, MergingDigest>>> {
+
+    abstract double compression();
+    abstract Builder<K> toBuilder();
+
+    static <K> Builder<K> builder() {
+      return new AutoValue_TDigestQuantiles_PerKeyDigest.Builder<K>()
+              .setCompression(100);
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder<K> {
+      abstract Builder<K> setCompression(double c);
+
+      abstract PerKeyDigest<K> build();
+    }
+
+    public PerKeyDigest<K> withCompression(double c) {
+      return toBuilder().setCompression(c).build();
+    }
+
+    @Override
+    public PCollection<KV<K, MergingDigest>> expand(PCollection<KV<K, Double>> input) {
+      return input.apply(
+              "Compute HyperLogLog Structure",
+              Combine.perKey(TDigestQuantilesFn.create(this.compression())));
+    }
+  }
+
+  /** Implements the {@link Combine.CombineFn} of {@link TDigestQuantiles} transforms. */
+  public static class TDigestQuantilesFn
+      extends Combine.CombineFn<Double, MergingDigest, MergingDigest> {
+
+    private final double compression;
+
+    private TDigestQuantilesFn(double compression) {
+      this.compression = compression;
+    }
+
+    /**
+     * Returns {@link TDigestQuantilesFn} combiner with the given compression factor.
+     *
+     * @param compression   the compression factor guarantees a relative error of at most
```

Comment:
Oh! This should be in the class javadoc. Also please explain more precisely what "relative error" means, as a mathematical inequality? E.g. if compression == 100, and I ask for 85% quantile - worst case, will it give me 82% to 88% quantile, or will it give me 0.97 to 1.03 * true 85%th quantile?

Reply:
I think I should rather delete this comment. I wrote it in the first version of TDigest based on what I read about TDigest but I can't find this information anymore in the paper. As it has been updated several time since, it might be possible that this information is no longer true.